### ISSUE TITLE
A method to drop zero columns of sparse matrix would be useful

### ISSUE THREAD

**Issue by EvgenyNekrasov** [scipy/scipy#6754]
It is in demand for machine learning problems, and it can be implemented efficiently for sparse matrices.

**Comment by perimosocordiae** [comment]
Assuming you can convert to CSC format, this isn't too hard to do in user code:
```python
nonzero_cols = np.diff(mat.indptr) != 0
new_indptr = mat.indptr[np.append(True, nonzero_cols)]
new_shape = (mat.shape[0], np.count_nonzero(nonzero_cols))
dropped = csc_matrix((mat.data, mat.indices, new_indptr), shape=new_shape)
```

**Comment by EvgenyNekrasov** [comment]
Yes, but the problem is that it does not work due to lack of memory when the amount of columns is really huge (e.g. 10^12), it is even impossible to convert matrix to CSC format. I am solving this problem by manipulating data in COO format, adding a following method to coo_matrix, dropping zero columns (majority), and converting to CSC or CSR format for downstream applications. 

```python
def drop_zero_columns(self):
    M, N = self.shape
    order = np.argsort(self.col)
    tmp = np.cumsum((np.diff(self.col[order]) > 0) * 1)
    self.col *= 0
    self.col[1:] += tmp
    self.row = self.row[order]
    self.data = self.data[order]
    N = int(tmp.max()) + 1
    self._shape = (M, N)
```

It looks like the easiest solution, but it forces a user to dig into SciPy internals. That's why I am asking this method.

**Comment by perimosocordiae** [comment]
Thanks for clarifying. I'm not sure how broadly useful such a method would be, so I'm interested to hear from others to get their opinion. You can email the scipy-dev mailing list to reach a broader audience; most users aren't subscribed to Github issues.

I would suggest a slightly different approach to your method:
```python
def drop_zero_columns(self):
    nz_cols, new_col = np.unique(self.col, return_inverse=True)
    self.col[:] = new_col
    self._shape = (self.shape[0], len(nz_cols))
```

**Comment by EvgenyNekrasov** [comment]
Thank you! Your implementation is great!

**Comment by rgommers** [comment]
Proposal of @EvgenyNekrasov on the mailing list: https://mail.scipy.org/pipermail/scipy-dev/2016-November/021594.html

+1 from me for implementing ``drop_zero_columns`` for all sparse matrix types.

**Comment by perimosocordiae** [comment]
Or perhaps drop_zero_axis?

On Nov 19, 2016 6:56 PM, "Ralf Gommers" <notifications@github.com> wrote:

Proposal of @EvgenyNekrasov <https://github.com/EvgenyNekrasov> on the
mailing list: https://mail.scipy.org/pipermail/scipy-dev/2016-
November/021594.html

+1 from me for implementing drop_zero_columns for all sparse matrix types.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/scipy/scipy/issues/6754#issuecomment-261695757>, or mute
the thread

**Comment by aki-nishimura** [comment]
I know this is an old thread, but +1 from me as well for this feature.

**Comment by rileyjmurray** [comment]
+1 for a "drop_zero_columns" feature. It would be especially helpful if the user could provide an iterable of column indices to be dropped (in my case there are some zero columns which I do not want removed).

**Comment by aronszanto** [comment]
Also an old thread, but I just came across it and it's exactly what I need. +1!

**Comment by XavierTolza** [comment]
Same for me. Would be useful
+1

**Comment by TomDeWeer** [comment]
+1, drop_zero_axis would be useful for me as well.

**Comment by perimosocordiae** [comment]
Trying to summarize a bit, we have a request for a new method, `drop_zero_axis`, which allows the user to specify either rows or columns to remove from a sparse matrix. We may also want a `drop_from_axis` method, which takes an iterable of indices along with the axis to remove them from.

I think these should be non-controversial, so I'll remove the `needs-decision` tag. Of course, we'll re-discuss when the PR comes in.

**Comment by akhmerov** [comment]
As an extra remark, the `drop_zero_axis` method should probably return the indices of nonzero columns/rows in the original matrix in addition to actually dropping the zero columns/rows.

**Comment by miketynes** [comment]
+1

**Comment by thecrazyfatguy** [comment]
+1

**Comment by lucascolley** [comment]
This has had a lot of requests by now so going to add a milestone, feel free to remove.

**Comment by h-vetinari** [comment]
More than a milestone, it needs someone who feels like tackling an implementation. ;-)

**Comment by ev-br** [comment]
Open from 2016, this feature request is about to graduate from an elementary school to a middle school :-). There were multiple requests to implement it over the years, so let me try bumping the milestone again and try to bring it to the radars of scipy.sparse regulars @perimosocordiae  @dschult -- is this something we'd want to have implemented still?

**Comment by perimosocordiae** [comment]
I'd still like to see it implemented.

**Comment by stefanv** [comment]
Accessible mailing list thread: https://mail.python.org/archives/list/scipy-dev@python.org/thread/JQCQJ6PT4FJNXBRJJO4DFAM6IO7DR53S/#VNTZAO4NH23YSNW6MD627UBY5TBCSPY5

**Comment by C-you-know** [comment]
Hi, this issue has been open for a while. I would like to contribute. Will submit a PR once it's ready. LMK if someone is already working on this. I will follow the above details for my implementation.

One question:  should drop_zero_axis return (array, kept_indices), or just the array with an optional return_indices keyword? Feature selection needs the indices, but a tuple return would be unusual for scipy.sparse methods.
